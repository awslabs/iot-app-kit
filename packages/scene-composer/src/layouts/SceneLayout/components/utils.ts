export enum Direction {
  Left = 'Left',
  Right = 'Right',
}
