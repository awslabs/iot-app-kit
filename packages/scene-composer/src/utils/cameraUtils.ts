import { Euler, Quaternion, Vector3 } from 'three';

import { ICameraComponentInternal } from '../store';
import { CameraSettings, ICameraBasics } from '../interfaces';
import { DEFAULT_CAMERA_POSITION, DEFAULT_CAMERA_SETTINGS } from '../common/constants';

export const getCameraSettings = (
  object3D: THREE.Object3D | undefined,
  cameraDetails: ICameraBasics | ICameraComponentInternal,
): CameraSettings => {
  const transform = object3D
    ? {
        position: object3D.getWorldPosition(new Vector3()).toArray(),
        rotation: new Vector3()
          .setFromEuler(new Euler().setFromQuaternion(object3D.getWorldQuaternion(new Quaternion())))
          .toArray(),
        scale: object3D.getWorldScale(new Vector3()).toArray(),
      }
    : {
        position: DEFAULT_CAMERA_POSITION,
      };

  if (cameraDetails) {
    return {
      cameraType: cameraDetails.cameraType,
      fov: cameraDetails.fov,
      far: cameraDetails.far,
      near: cameraDetails.near,
      zoom: cameraDetails.zoom,
      transform,
    };
  }

  return {
    ...DEFAULT_CAMERA_SETTINGS,
    cameraType: 'Perspective',
    transform,
  };
};
