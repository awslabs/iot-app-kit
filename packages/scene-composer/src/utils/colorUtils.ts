/**
 * This method uses a regular expression (`hexRegex`) to validate a hex color code.
 * The regex checks if the hex code starts with a "#" symbol, followed by either a
 * 6-digit combination of characters from A-F, a-f, and 0-9.
 * The `test` method is then used to validate the `hexCode` against the regex pattern.
 * @param hexCode
 * @returns
 */
export const isValidHexCode = (hexCode: string): boolean => {
  const hexRegex = /^#([A-Fa-f0-9]{6})$/;
  return hexRegex.test(hexCode);
};
