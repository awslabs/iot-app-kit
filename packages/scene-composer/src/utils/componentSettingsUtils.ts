import { RootState } from '../store';
import { KnownSceneProperty, IComponentSettings, KnownComponentType } from '../interfaces';
import { DEFAULT_OVERLAY_GLOBAL_SETTINGS, DEFAULT_TAG_GLOBAL_SETTINGS } from '../common/constants';

export const componentSettingsSelector = (state: RootState, componentType: KnownComponentType): IComponentSettings => {
  const settings = state.getSceneProperty(KnownSceneProperty.ComponentSettings)?.[componentType];
  if (settings) {
    return settings;
  } else {
    // When the settings is not in scene file, return the default setting for each component type.
    switch (componentType) {
      case KnownComponentType.Tag:
        return DEFAULT_TAG_GLOBAL_SETTINGS;
      case KnownComponentType.DataOverlay:
        return DEFAULT_OVERLAY_GLOBAL_SETTINGS;
      default:
        return {};
    }
  }
};
