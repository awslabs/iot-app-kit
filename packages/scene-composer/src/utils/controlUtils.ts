import { MapControls, OrbitControls } from '../three/OrbitControls';
import { PointerLockControls } from '../three/PointerLockControls';
import { CameraControlsType } from '../interfaces';
import { CameraControlImpl } from '../store/internalInterfaces';
import { Pan } from '../assets/auto-gen/icons';

export const isPointerLockControl = (controlType: CameraControlsType): boolean => {
  return controlType === 'pointerLock' ? true : false;
};

export const isOrbitControl = (controlType: CameraControlsType): boolean => {
  return controlType === 'orbit' ? true : false;
};

export const isPanControl = (controlType: CameraControlsType): boolean => {
  return controlType === 'pan' ? true : false;
};

export const isImmersiveControl = (controlType: CameraControlsType): boolean => {
  return controlType === 'immersive' ? true : false;
};

export const isOrbitOrPanControl = (controlType: CameraControlsType): boolean => {
  return isOrbitControl(controlType) || isPanControl(controlType) ? true : false;
};

export const isPointerLockControlImpl = (controler: CameraControlImpl): boolean => {
  return controler instanceof PointerLockControls ? true : false;
};

// note that MapControls is an extension of OrbitControls soMapControls will return true here also
export const isOrbitControlImpl = (controler: CameraControlImpl): boolean => {
  return controler instanceof OrbitControls ? true : false;
};

export const isPanControlImpl = (controler: CameraControlImpl): boolean => {
  return controler instanceof MapControls ? true : false;
};

export const isOrbitOrPanControlImpl = (controler: CameraControlImpl): boolean => {
  return isOrbitControlImpl(controler) || isPanControlImpl(controler) ? true : false;
};
