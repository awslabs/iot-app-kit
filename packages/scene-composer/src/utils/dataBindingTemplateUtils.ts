import { cloneDeep, pick } from 'lodash';

import {
  DEFAULT_DATA_BINDING_TEMPLATE_COMPONENT_NAME,
  DEFAULT_DATA_BINDING_TEMPLATE_ENTITY_ID,
  DataBindingLabelKeys,
} from '../common/constants';
import {
  IDataBindingConfig,
  IDataBindingTemplate,
  IDataFieldOption,
  IValueDataBinding,
  KnownSceneProperty,
} from '../interfaces';
import { RootState } from '../store';

/**
 * Data binding templates will be stored as ${my-value} in IValueDataBinding
 */
const dataBindingTemplateRegExp = /^\$\{([\s\S]+)\}$/;

export const isDataBindingTemplate = (item?: string) => (item ? dataBindingTemplateRegExp.test(item) : false);

export const decorateDataBindingTemplate = (item: string) => '${' + item + '}';

export const undecorateDataBindingTemplate = (item: string) => item.match(dataBindingTemplateRegExp)?.[1] ?? item;

export const dataBindingConfigSelector = (state: RootState): IDataBindingConfig => {
  const dataBindingConfig: IDataBindingConfig =
    cloneDeep(state.getSceneProperty(KnownSceneProperty.DataBindingConfig)) ?? {};
  dataBindingConfig.fieldMapping = dataBindingConfig.fieldMapping ?? {};
  dataBindingConfig.fieldMapping.entityId = dataBindingConfig.fieldMapping.entityId ?? [
    DEFAULT_DATA_BINDING_TEMPLATE_ENTITY_ID,
  ];
  dataBindingConfig.fieldMapping.componentName = dataBindingConfig.fieldMapping.componentName ?? [
    DEFAULT_DATA_BINDING_TEMPLATE_COMPONENT_NAME,
  ];
  return dataBindingConfig;
};

/**
 * Create selection options for data binding templates, these options will be shown on top of existing options. E.g. ${sel_entity} will
 * be shown before all entity IDs
 * @param fieldName field name, e.g. entityId, componnentName
 * @param dataBindingConfig current data binding configuration
 * @param lastSelectedOption only if the last selected option is data binding template, then the current options will include data binding template
 * @returns
 */
export const createDataBindingTemplateOptions = (
  fieldName: string,
  dataBindingConfig?: IDataBindingConfig,
  lastSelectedOption?: string,
): IDataFieldOption[] => {
  if (!dataBindingConfig) {
    return [];
  }

  if (lastSelectedOption && !isDataBindingTemplate(lastSelectedOption)) {
    return [];
  }

  return dataBindingConfig.fieldMapping[fieldName]
    .filter((templateName) => !!dataBindingConfig.template?.[templateName])
    .map(decorateDataBindingTemplate)
    .map((templateName) => ({ value: templateName, label: templateName }));
};

// Undecorate data binding (from "${XX} to XXX").
//
// Example of dataBinding
// {
//   "dataBindingContext": {
//     "entityId": "${sel_entity}",
//     "componentName": "${sel_comp}",
//     "propertyName": "temperature"
//    }
// }
//
// Example of dataBindingTemplate
// {
//   "sel_entity": "room1",
//   "sel_comp": "temperatureSensor2"
// },
export function applyDataBindingTemplate(
  dataBinding?: IValueDataBinding,
  dataBindingTemplate?: IDataBindingTemplate,
): Record<string, string> {
  if (!dataBinding?.dataBindingContext || !dataBindingTemplate) {
    return (dataBinding?.dataBindingContext ?? {}) as Record<string, string>;
  }

  const dataBindingContext = cloneDeep(dataBinding.dataBindingContext) as {};
  const bindingKeys = Object.keys(dataBinding.dataBindingContext as any);
  bindingKeys.forEach((bindingKey) => {
    const undecoratedBindingValue = undecorateDataBindingTemplate(dataBindingContext[bindingKey]);
    dataBindingContext[bindingKey] = dataBindingTemplate[undecoratedBindingValue] ?? dataBindingContext[bindingKey];
  });
  return dataBindingContext;
}

/**
 * We extract entityId from the data binding object.
 * @param dataBinding we send data binding object
 * @returns
 */
export const extractEntityId = (dataBinding: IValueDataBinding): Record<string, string> => {
  const contextData = dataBinding.dataBindingContext ?? {};
  return pick(contextData, [DataBindingLabelKeys.entityId]);
};
