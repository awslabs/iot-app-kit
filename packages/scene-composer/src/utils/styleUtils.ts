import {
  colorChartsYellow700,
  colorForegroundControlDefault,
  colorBackgroundNotificationBlue,
  colorTextNotificationDefault,
  colorBackgroundNotificationRed,
  colorBorderButtonNormalDefault,
} from '@awsui/design-tokens';

const designTokenRegex = /^var\(([0-9a-zA-Z-]+),\s*(#[0-9a-z]+)\)$/;

/**
 * extract css hex color value from design token.
 *
 * @param designToken the design token
 * @returns the hex color extracted from the design token
 */
export function hexColorFromDesignToken(designToken: string): string {
  const match = designToken.match(designTokenRegex);
  if (!match) return '#000000';
  const varName = match[1];
  const defaultValue = match[2];

  return getComputedStyle(document.body).getPropertyValue(varName) || defaultValue;
}

// Prefer the values to be stripped out of the style provided by AWSUI. Default values are backups of the current values.
export const colors = {
  infoBlue: hexColorFromDesignToken(colorBackgroundNotificationBlue || '#0073bb'),
  warnYellow: hexColorFromDesignToken(colorChartsYellow700 || '#dfb52c'),
  errorRed: hexColorFromDesignToken(colorBackgroundNotificationRed || '#d13212'),
  selectedOrange: '#ec7211', // cannot use colorTextAccent since it is blue for external AWSUI lib
  infoRingWhite: hexColorFromDesignToken(colorTextNotificationDefault || '#fafafa'),
  symbolWhite: hexColorFromDesignToken(colorForegroundControlDefault || '#fafafa'),
  containerBorderWhite: hexColorFromDesignToken(colorBorderButtonNormalDefault || '#fafafa'),
};
