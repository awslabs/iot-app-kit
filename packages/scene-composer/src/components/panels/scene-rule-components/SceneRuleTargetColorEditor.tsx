import React, { FC, useCallback, useState } from 'react';
import { ChromePicker } from 'react-color';
import { Input, Grid } from '@awsui/components-react';

import { colors } from '../../../utils/styleUtils';

interface ColorSwatchProps {
  backgroundColor: string;
  borderColor: string;
  onClick: () => void;
}

const ColorSwatch: FC<ColorSwatchProps> = ({ backgroundColor, borderColor, ...props }) => {
  return (
    <div
      style={{
        backgroundColor,
        borderColor,
        width: '100%',
        height: '30px',
        borderStyle: 'solid',
        borderWidth: '1px',
        cursor: 'pointer',
      }}
      {...props}
    />
  );
};

interface ISceneRuleTargetColorEditorProps {
  targetValue: string;
  onChange: (target: string) => void;
}

export const SceneRuleTargetColorEditor: React.FC<ISceneRuleTargetColorEditorProps> = ({
  targetValue,
  onChange,
}: ISceneRuleTargetColorEditorProps) => {
  const [showColorPicker, setShowColorPicker] = useState(false);

  const toggleColorPicker = useCallback(() => setShowColorPicker(!showColorPicker), [showColorPicker]);

  return (
    <Grid gridDefinition={[{ colspan: 9 }, { colspan: 3 }]}>
      <Input disabled={true} value={targetValue} />
      <ColorSwatch
        data-testid='color-swatch'
        onClick={toggleColorPicker}
        backgroundColor={targetValue}
        borderColor={colors.containerBorderWhite}
      />
      {showColorPicker && (
        <ChromePicker
          disableAlpha
          color={targetValue}
          onChangeComplete={(newColor: any) => {
            onChange(newColor.hex);
          }}
        />
      )}
    </Grid>
  );
};
