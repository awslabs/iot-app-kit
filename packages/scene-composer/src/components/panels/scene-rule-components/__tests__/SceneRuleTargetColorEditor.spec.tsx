import React from 'react';
import { render } from '@testing-library/react';
import wrapper from '@awsui/components-react/test-utils/dom';

import { SceneRuleTargetColorEditor } from '../SceneRuleTargetColorEditor';
import { colors } from '../../../../utils/styleUtils';

describe('SceneRuleTargetColorEditor', () => {
  it('should open the chrome-picker on click', () => {
    const { container } = render(<SceneRuleTargetColorEditor targetValue={colors.infoBlue} onChange={jest.fn()} />);
    const polarisWrapper = wrapper(container);
    const colorSwatch = polarisWrapper.find('[data-testid="color-swatch"]');
    colorSwatch!.click();

    const colorPicker = polarisWrapper.findByClassName('chrome-picker');
    expect(colorPicker).not.toBeNull();
  });
});
