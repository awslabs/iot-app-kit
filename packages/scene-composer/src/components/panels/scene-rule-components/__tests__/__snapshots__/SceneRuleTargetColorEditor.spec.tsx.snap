// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SceneRuleTargetColorEditor should render as expected 1`] = `
<div>
  <div
    data-mocked="Grid"
    griddefinition="[{\\"colspan\\":9},{\\"colspan\\":3}]"
  >
    <div
      data-mocked="Input"
      disabled=""
      value="#000000"
    />
    <div
      data-testid="color-swatch"
      style="background-color: rgb(0, 0, 0); border-color: #000000; width: 100%; height: 30px; border-style: solid; border-width: 1px; cursor: pointer;"
    />
  </div>
</div>
`;

exports[`SceneRuleTargetColorEditor should toggle color picker on click 1`] = `
<div>
  <div
    data-mocked="Grid"
    griddefinition="[{\\"colspan\\":9},{\\"colspan\\":3}]"
  >
    <div
      data-mocked="Input"
      disabled=""
      value="#000000"
    />
    <div
      data-testid="color-swatch"
      style="background-color: rgb(0, 0, 0); border-color: #000000; width: 100%; height: 30px; border-style: solid; border-width: 1px; cursor: pointer;"
    />
    <div
      color="#000000"
      data-mocked="ChromePicker"
    />
  </div>
</div>
`;
